proxy_http.go

```diff
@@ -32,14 +33,24 @@ func (opts *Opts) applyHTTPDefaults() {
 }
 
 // Handle implements the interface Proxy
-func (proxy *proxy) Handle(ctx context.Context, downstream net.Conn) error {
+func (proxy *proxy) Handle(ctx context.Context, downstreamIn io.Reader, downstream net.Conn) error {
+	return proxy.handle(ctx, downstreamIn, downstream, nil)
+}
+
+func (proxy *proxy) handle(ctx context.Context, downstreamIn io.Reader, downstream net.Conn, upstream net.Conn) error {
 	defer func() {
 		if closeErr := downstream.Close(); closeErr != nil {
 			log.Tracef("Error closing downstream connection: %s", closeErr)
 		}
 	}()
 
-	downstreamBuffered := bufio.NewReader(downstream)
+	var downstreamBuffered *bufio.Reader
+	switch r := downstreamIn.(type) {
```

Comment:
Ah, cool, I didn't realize the bufio.NewReader already has this logic :) We don't actually know what's passed in, because Handle is a public method. Still, bufio.NewReader takes care of this for us, so I've removed the logic.